endpoint/endpoint_diagnoseComp/endpoint_diagnose.c

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (C) 2019-2020 BiiLabs Co., Ltd. and Contributors
+ * All Rights Reserved.
+ * This is free software; you can redistribute it and/or modify it under the
+ * terms of the MIT license. A copy of the license can be found in the file
+ * "LICENSE" at the root of this distribution.
+ */
+
+#include "endpoint_diagnose.h"
+
+#include "legato.h"
+#include "le_gnss_interface.h"
+#include "le_mrc_interface.h"
+
+COMPONENT_INIT {
+
+  char time_str[30];
+  size_t copied_bytes = 0;
+  int32_t latitude = 0, longitude = 0, accuracy = 0;
+  uint32_t quality = 0;
+  le_gnss_SampleRef_t position_sample;
+  le_gnss_State_t gnss_state;
+  le_result_t result;
+
+  // GPS state checking and activation  
+  gnss_state = le_gnss_GetState();
+  if (gnss_state != LE_GNSS_STATE_ACTIVE) {
+    le_gnss_Start();
+    le_gnss_Enable();
+  }
+
+  while(1) {
+    // Signal strength
+    result = le_mrc_GetSignalQual(&quality);
+    if (result == LE_OK) {
+      LE_INFO("Signal strength: %d\n", quality);
+    } else {
+      LE_ERROR("Fail to get signal strength\n");
+    }
+
+    // GPS
+    position_sample = le_gnss_GetLastSampleRef();
+    result = le_gnss_GetLocation(position_sample, &latitude, &longitude, &accuracy);
+    if (result == LE_OK) {
```

Comment:
Rewrite with switch-case for readability and potential performance gain.

Reply:
Done.